pkg/controllers/cassandra/pilot/pilot_test.go

```diff
@@ -1,45 +1,56 @@
 package pilot_test
 
 import (
+	"fmt"
 	"testing"
 
+	"k8s.io/api/apps/v1beta1"
 	"k8s.io/api/core/v1"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 
+	"github.com/kr/pretty"
+
 	"github.com/jetstack/navigator/internal/test/unit/framework"
 	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
 	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/cassandra/nodepool"
 	"github.com/jetstack/navigator/pkg/controllers/cassandra/pilot"
 	casstesting "github.com/jetstack/navigator/pkg/controllers/cassandra/testing"
-	"github.com/jetstack/navigator/pkg/controllers/cassandra/util"
 )
 
-func clusterPod(cluster *v1alpha1.CassandraCluster, name string) *v1.Pod {
-	pod := &v1.Pod{}
-	pod.SetName(name)
-	pod.SetNamespace(cluster.GetNamespace())
-	pod.SetOwnerReferences(
-		[]metav1.OwnerReference{
-			util.NewControllerRef(cluster),
+func clusterPod(cluster *v1alpha1.CassandraCluster, set *v1beta1.StatefulSet, index int32) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%d", set.Name, index),
+			Namespace: set.Namespace,
+			Labels: map[string]string{
+				v1alpha1.CassandraClusterNameLabel:  cluster.Name,
+				v1alpha1.CassandraNodePoolNameLabel: set.Labels[v1alpha1.CassandraNodePoolNameLabel],
+			},
 		},
-	)
-	return pod
+	}
 }
 
 func TestPilotSync(t *testing.T) {
 	cluster1 := casstesting.ClusterForTest()
-	cluster1pod1 := clusterPod(cluster1, "c1p1")
-	cluster1pod2 := clusterPod(cluster1, "c1p2")
-	cluster1pilot1 := pilot.PilotForCluster(cluster1, cluster1pod1)
-	cluster1pilot1foreign := cluster1pilot1.DeepCopy()
-	cluster1pilot1foreign.SetOwnerReferences([]metav1.OwnerReference{})
+	cluster1np1 := &cluster1.Spec.NodePools[0]
+	cluster1np1set := nodepool.StatefulSetForCluster(cluster1, cluster1np1)
+	cluster1np1set.Status.CurrentReplicas = 1
+	cluster1np1pod0 := clusterPod(cluster1, cluster1np1set, 0)
+	cluster1np1pod1 := clusterPod(cluster1, cluster1np1set, 1)
+	cluster1np1pilot0 := pilot.PilotForCluster(cluster1, cluster1np1pod0)
+	cluster1np1pilot1 := pilot.PilotForCluster(cluster1, cluster1np1pod1)
+	cluster1np1pilot0foreign := cluster1np1pilot0.DeepCopy()
+	cluster1np1pilot0foreign.SetOwnerReferences([]metav1.OwnerReference{})
 
 	cluster2 := casstesting.ClusterForTest()
 	cluster2.SetName("cluster2")
 	cluster2.SetUID("uid2")
-	cluster2pod1 := clusterPod(cluster2, "c2p1")
+	cluster2np1 := &cluster2.Spec.NodePools[0]
+	cluster2np1set := nodepool.StatefulSetForCluster(cluster2, cluster2np1)
+	cluster2np1pod0 := clusterPod(cluster2, cluster2np1set, 0)
```

Comment:
It'd be great if we can avoid such a huge stack of objects declared here. It's really difficult when looking at a particular test case to see what's being tested (i.e. the inputs and outputs).

Not sure if there's an easy way to do this, but right now I am definitely struggling to understand all this.

I managed to get around this in cert-manager by accepting a pointer to the fixture itself as part of a 'PreFn': https://github.com/jetstack/cert-manager/blob/master/pkg/issuer/acme/http/ingress_test.go#L21-L54

If not, comments throughout all of this may help make this more readable 😄

Reply:
With the revised implementation I found I could remove many of these.
I tried to make the variable names self-explanatory. 
Not sure what more I can add with comments.